flow/protocolversion/protocol_version.py

```diff
@@ -0,0 +1,345 @@
+
+    def set_min_invalid_version(self, version: int):
+        self._min_invalid_version = version
+
+    def set_left_most_check(self, version: int):
+        self._left_most_check = version
+
+    def set_lsb_mask(self, version: int):
+        self._lsb_mask = version
+
+    @property
+    def default_version(self):
+        return self._default_version
+
+    @property
+    def future_version(self):
+        return self._future_version
+
+    @property
+    def min_compatible_version(self):
+        return self._min_compatibile_version
+
+    @property
+    def min_invalid_version(self):
+        return self._min_invalid_version
+
+    @property
+    def left_most_check(self):
+        return self._left_most_check
+
+    @property
+    def lsb_mask(self):
+        return self._lsb_mask
+
+    def add_feature(self, version: int, feature: str):
+        if version not in self._features:
+            self._features[version] = []
+        self._features[version].append(feature)
+
+        return self
+
+    @property
+    def features(self):
+        return self._features
+
+
+def _decapitate(text: str, prefix: str) -> str:
+    if text.startswith(prefix):
+        return text[len(prefix) :]
+    return text
+
+
+def _tail_chop(text: str, postfix: str) -> str:
+    if text.endswith(postfix):
+        return text[: len(text) - len(postfix)]
+    return text
+
+
+class ProtocolVersionSerializerBase(abc.ABC):
+    @abc.abstractmethod
+    def _load(self, stream: io.TextIOWrapper) -> ProtocolVersion:
+        raise NotImplementedError()
+
+    def load(self, stream: io.TextIOWrapper) -> ProtocolVersion:
+        return self._load(stream)
+
+    @abc.abstractmethod
+    def _save(self, protocol_version: ProtocolVersion, stream: io.TextIOWrapper):
+        raise NotImplementedError()
+
+    def save(self, protocol_version: ProtocolVersion, stream: io.TextIOWrapper):
+        return self._save(protocol_version, stream)
+
+
+class CMakeProtocolVersionSerializer(ProtocolVersionSerializerBase):
+    SPECIAL_FIELDS = {
+        "DEFAULT_VERSION": ProtocolVersion.set_default_version,
+        "FUTURE_VERSION": ProtocolVersion.set_future_version,
+        "MIN_COMPATIBLE_VERSION": ProtocolVersion.set_min_compatible_version,
+        "MIN_INVALID_VERSION": ProtocolVersion.set_min_invalid_version,
+        "LEFT_MOST_CHECK": ProtocolVersion.set_left_most_check,
+        "LSB_MASK": ProtocolVersion.set_lsb_mask,
+    }
+
+    def _decode_version(self, encoded: str) -> int:
+        """Decode version like 0x0FDB00B073000000LL into decimal value"""
+        return int(_tail_chop(encoded, "LL"), 16)
+
+    def _load(self, stream: io.TextIOWrapper) -> ProtocolVersion:
+        protocol_version_cmake_regex = re.compile(
+            r"set\((?P<feature>[^\s]+)\s+\"(?P<version>[^\"]+)\"\)"
+        )
+        protocol_version = ProtocolVersion()
+        for line in stream:
+            match = protocol_version_cmake_regex.search(line)
+            if not match:
```

Comment:
I am not sure if this would pollute lots of warnings since in the `ProtocolVersions.cmake` there are plenty of comments. I suppose we can move to YAML format later so we do not need to use regexes to match the text.

Reply:
I think we can ignore `cmake` comments, e.g., `# Protocol Versions.`. But I do think it's important to emit warnings for anything else that's not matching.